### ISSUE TITLE
Some keys don't show up on Input-History

### ISSUE THREAD

**Issue by RoyZ7** [univrsal/input-overlay#114]
Hi friend, 

The keys: Insert, Delete, Home, End, PageUp, PageDown Don't show up on Input-History using plain text.

The last version they've been working at was version 4.0 and they only showed the hex values, not the key names. 

Thank you so much for this plug-in 🙏

**Comment by Viginox** [comment]
Could you also share the version of the plugin and which operating system you are using?

And are you using a custom key name config file?

I think a hotfix would be to use a key name config file and add those keys there

```
2_<hex>=INSERT
2_<hex>=DELETE
2_<hex>=HOME
2_<hex>=PAGE_UP
2_<hex>=PAGE_DOWN
```
Replace `<hey>` with the value for that key.
(You can also rename the key to any name you want after the `=` )
And don't forget to tick the box "Use builtin names if undefined in file"

**Comment by RoyZ7** [comment]
Thanks, I'm using version 4.8 of the plugin and the currently recent OBS version 25.0.8 on a Windows 10 64bit system.

I've tried your suggested hotfix but it didn't work for me, maybe I put the code in the wrong place or used the wrong syntax?
 
I'm using one of the submitted config files, I'm attaching it here for your review after I've added the code you've suggested to it.

I didn't forget to tick the box "Use builtin names if undefined in file" :)

[minimal.txt](https://github.com/univrsal/input-overlay/files/4733208/minimal.txt)

Thank you!

**Comment by Viginox** [comment]
It seems to be the current version (except the upcoming v5).

I'm only running it on Linux x64, so I cannot confirm if it is working on Windows too.
I had some issues with the "wrong" keys showing up for the keymap I was using and I could fix it with the custom config file.

The file, you attached to your last comment is **not** a config file for *input-history*! It is **only** used for *input-overlay*. 

## Hotfix (Guided)
So here again, a more detailed version of how to apply that hotfix:

1. create the config file somewhere on your system. Let's call it `input-history-fix.ini` and place it on the Desktop (`$HOME/Desktop`) 
    * total filepath: `$HOME/Desktop/input-history-fix.ini` 
    --> I will refer to that from now on as **config file**
1. Open the **config file** and OBS
1. Copy the text from my first comment to the **config file**. Each line must start with a `2`!
    * Alternatively copy the content from the next section to the file and continue with **step 8**
1. Go to OBS and press the `INSERT` key
1. Watch the hex code appear in the input-history overlay in OBS
1. Replace `<hex>` in the first line of the **config file** with that hex code (i.e. `0x0015`)
1. Repeat for all keys
1. Save the **config file**
1. Open the Settings for the *input-history* source in OBS
1. Select the **config file** in the Setting `Key name config`
1. Enable `Use bulitin names if undefined in file`
1. Klick `OK` to save the settings

This should override the default values and you should see the text, you set in the **config file** for that key, when you press it.

If everything works out as expected, I recommend moving the **config key** to a designated folder and updating the *input-history* config accordingly.

I hope this helps!

## File
According to your "minimal.txt" file, the **config file** should look like this:
```
# <-- this is a comment
# You can use everything as keynames.
# Remove # from the next line to test this
#2_0x0E52=My name is insert key <3
# Add a # to the next line to test the line above!
2_0x0E52=INSERT
2_0x0E53=DELETE
2_0x0E47=HOME
2_0x0E4F=END
2_0x0E49=PAGE UP
2_0x0E51=PAGE DOWN
```

**Comment by RoyZ7** [comment]
Thank you for trying to help (:

I've tried it your fix in any case but it doesn't work as the problem is different, in version 4.8 (at least on windows OS) the hex codes do not show up at all with those 6 keys, so there is no code to change to another text by a config file hotfix. these keys DO show the hex value all the way up to version 4.0 but then the feature to use "key icons" which I'd like to have, didn't exist yet, so it's not available. 

The config file I was using was actually for input-history, when you use the "key icons" feature of it rather than "plain text". that's the way you can show graphic icons for the keys that are pressed.

It's something broken in the code, Thanks for trying again and the detailed hotfix suggestion  (:

**Comment by Viginox** [comment]
I just set up a Windows VM and tried it for myself.
It seems, it is working really strange under windows ... Apparently there is something wrong, but I'm not that much into the code to figure out what exactly.

If I come across a solution for that, I'll let you know.

Until then, I'm sorry, I couldn't be more of a help to you =/

**Comment by univrsal** [comment]
> # <-- this is a comment
> # You can use everything as keynames.
> # Remove # from the next line to test this
> #2_0x0E52=My name is insert key <3
> # Add a # to the next line to test the line above!
> 2_0x0E52=INSERT
> 2_0x0E53=DELETE
> 2_0x0E47=HOME
> 2_0x0E4F=END
> 2_0x0E49=PAGE UP
> 2_0x0E51=PAGE DOWN

That will most likely not work since I changed the config system in 4.8. The prefix should be '1_' but it also might be an issue with input history

**Comment by Viginox** [comment]
That's interesting because those settings are working for me on Manjaro Linux with version 4.8 and also on Windows 10 [VM] (on some keys).

In addition, the Wiki still says to use values with `2_` in front.
https://github.com/univrsal/input-overlay/wiki/Creating-presets#input-history

###########

If you want some help, looking into this problem, could you give me a hint on where to look for the issue?

**Comment by RoyZ7** [comment]
Thank you @Viginox for your help and for keep trying!

and Thank you @univrsal as well, I think it's something to do with the implementation of the "libUIOHook" code right? I wish I could dig in and help as well but unfortunately I don't know C++, Really hoping this issue will be solved 🙏

**Comment by univrsal** [comment]
> That's interesting because those settings are working for me on Manjaro Linux with version 4.8 and also on Windows 10 [VM] (on some keys).
> 
> In addition, the Wiki still says to use values with `2_` in front.
> https://github.com/univrsal/input-overlay/wiki/Creating-presets#input-history
> 
> ###########
> 
> If you want some help, looking into this problem, could you give me a hint on where to look for the issue?

The wiki might not be up to date, but honestly I don't even know anymore. I've spend most time on the master branch which wildly differs from the latest release.

> Thank you @Viginox for your help and for keep trying!
> 
> and Thank you @univrsal as well, I think it's something to do with the implementation of the "libUIOHook" code right? I wish I could dig in and help as well but unfortunately I don't know C++, Really hoping this issue will be solved

The keys should work, it's just that libuiohook seems to switch up the Insert, Delete, Home, End, PageUp, PageDown keys with the ones from the numpad
![image](https://user-images.githubusercontent.com/8353672/84023382-85402f80-a977-11ea-9d73-99a65fc1e40d.png)

**Comment by RoyZ7** [comment]
@univrsal Thank you, yes I know those work, it's just that it would be better to use the numpad for numbers and those 6 keys with their dedicated keys, that is if you want to show those keys on screen and not toggle the numlock key every time. maybe we could ask the libuiohook developer for advice? Thanks a lot.

**Comment by univrsal** [comment]
Let me rephrase that, both versions of those keys should work, the keycodes are switched up though. Which means that the ones that are labled ``_KP`` are actually the normal ones and the ones that are labeled normal are actually the ones on the keypad. At least that's how it's been for most other users.

**Comment by RoyZ7** [comment]
I see.. Thanks, a bit confused :) when I press any of those 6 keys while not on the keypad, using "Plain text" I don't get any hex value on screen.. so is there any way I could make them show the keys when I press them?

**Comment by univrsal** [comment]
Well if what I say is correct, you should just follow what @Viginox said but instead of using the normal numpad codes you'd use the keypad codes:
```
# Key names
1_0xEE52=INSERT
1_0xEE53=DELETE
1_0xEE47=HOME
1_0xEE4F=END
1_0xEE49=PAGE UP
1_0xEE51=PAGE DOWN

```
Mind the empty line at the end, sometimes people experienced issues if it's missing

**Comment by Viginox** [comment]
After all that info, I just did some testing on my Windows VM.

I compiled `libuiohook` and used the demo project to get the keycodes for my keys.
They are the same as in the post above, but when using them in the config file, I had some issues with the `1_` prefix because it was just not working ...
After some testing with the latest release version, I figured out that the following config file was working for me:

```
# Key names
2_0xEE53=DELETE
2_0xEE52=INSERT
2_0xEE4F=END
2_0xEE47=HOME
2_0xEE49=PAGE UP
2_0xEE51=PAGE DOWN

```
So I don't know why, but the prefix `1_` is just not working in all my tests.

Those values did not mess up the names for the numpad.
Please keep in mind, those values were tested on a Windows VM with a bluetooth numpad.

## Possible Feature
After all those troubles, figuring out the correct key codes, how about a feature to get the correct codes for a key?

**Comment by univrsal** [comment]
The demo programs are usually used for finding out the keycodes, but I technically intended the config creation tool to take care of that, whenever I get around to finishing that. Also you don't have to go through the trouble of building libuiohook anymore since it recently got automatic builds via [github actions](https://github.com/kwhat/libuiohook/actions/runs/117016774)

**Comment by univrsal** [comment]
See #122
